crates/revm/src/evm.rs

```diff
@@ -68,133 +67,95 @@ impl<DB: Database + DatabaseCommit> EVM<DB> {
+            &mut NoOpInspector,
+        )
+        .preverify_transaction()
     }
 
     /// Skip preverification steps and execute transaction
     /// without writing to DB, return change state.
     pub fn transact_preverified_ref(&self) -> EVMResult<DB::Error> {
-        if let Some(db) = self.db.as_ref() {
-            evm_inner::<_, false>(
-                &mut self.env.clone(),
-                &mut WrapDatabaseRef(db),
-                &mut NoOpInspector,
-            )
-            .transact_preverified()
-        } else {
-            panic!("Database needs to be set");
-        }
+        evm_inner::<_, false>(
+            &mut self.env.clone(),
+            &mut WrapDatabaseRef(&self.db),
+            &mut NoOpInspector,
+        )
+        .transact_preverified()
     }
 
     /// Execute transaction without writing to DB, return change state.
     pub fn transact_ref(&self) -> EVMResult<DB::Error> {
-        if let Some(db) = self.db.as_ref() {
-            evm_inner::<_, false>(
-                &mut self.env.clone(),
-                &mut WrapDatabaseRef(db),
-                &mut NoOpInspector,
-            )
-            .transact()
-        } else {
-            panic!("Database needs to be set");
-        }
+        evm_inner::<_, false>(
+            &mut self.env.clone(),
+            &mut WrapDatabaseRef(&self.db),
+            &mut NoOpInspector,
+        )
+        .transact()
     }
 
     /// Execute transaction with given inspector, without wring to DB. Return change state.
     pub fn inspect_ref<I: Inspector<WrapDatabaseRef<&'a DB>>>(
         &'a self,
         mut inspector: I,
     ) -> EVMResult<DB::Error> {
-        if let Some(db) = self.db.as_ref() {
-            evm_inner::<_, true>(
-                &mut self.env.clone(),
-                &mut WrapDatabaseRef(db),
-                &mut inspector,
-            )
-            .transact()
-        } else {
-            panic!("Database needs to be set");
-        }
+        evm_inner::<_, true>(
+            &mut self.env.clone(),
+            &mut WrapDatabaseRef(&self.db),
+            &mut inspector,
+        )
+        .transact()
     }
 }
 
 impl<DB> EVM<DB> {
     /// Creates a new [EVM] instance with the default environment,
-    pub fn new() -> Self {
-        Self::with_env(Default::default())
+    pub fn new(db: DB) -> Self {
+        Self::with_env(Default::default(), db)
     }
 
     /// Creates a new [EVM] instance with the given environment.
-    pub fn with_env(env: Env) -> Self {
-        Self { env, db: None }
-    }
-
-    pub fn database(&mut self, db: DB) {
-        self.db = Some(db);
-    }
-
-    pub fn db(&mut self) -> Option<&mut DB> {
-        self.db.as_mut()
+    pub fn with_env(env: Env, db: DB) -> Self {
+        Self { env, db }
     }
 
-    pub fn take_db(&mut self) -> DB {
-        core::mem::take(&mut self.db).unwrap()
+    pub fn db(&mut self) -> &mut DB {
+        &mut self.db
     }
 }
```

Comment:
Can you revert this so as to not break functions

`pub fn new() -> Self {` should init with `EmptyDB` this would remove a need to have `Option` and would make code nicer